### PR TITLE
issue: 4176956: move the text handling to the analyzers classes

### PR DESCRIPTION
**Pull request by Miryam-Schwartz**
## What
_Move text and dataframes handling from main to the analyzers classes._ 

## Why ?
https://redmine.mellanox.com/issues/4176956
_For the sake of code aesthetics, readability, and encapsulation._

## How ?
_Some sub-classes of BaseAnalyzer class, override the implementation of analysis method, to add their own logic (handle text and dataframes)._

## Testing ?
_Tested manually with various inputs to ensure the output file remains consistent._

## Special triggers
_Use the following phrases as comments to trigger different runs_

* ```bot:retest``` rerun Jenkins CI (to rerun GitHub CI, use "Checks" tab on PR page and rerun _all_ jobs)
* ```bot:upgrade``` run additional update tests

### CODE REVIEW

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py [hunk 1]
```diff
@@ -369,7 +369,8 @@ def create_analyzer(
         end = time.perf_counter()
         log.LOGGER.debug(f"Took {end-start:.3f} to load the parsed data")
 
-        all_images_outputs_and_title = ufm_top_analyzer.full_analysis()
+        all_images_outputs_and_title, dataframes_for_pdf, lists_to_add = \
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: lists_to_add -> texts_for_pds?

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py [hunk 2]
```diff
@@ -57,6 +57,8 @@ def __init__(self, dest_image_path):
     def _save_data_based_on_timestamp(
         self, data_to_plot, x_label, y_label, title, large_sample=False
     ):
+        if len(data_to_plot) == 0:
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py:60]: I think data_to_plot is a panda dataframe; if so, this should be data_to_plot.empty() (Since len will count the headers)

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py [hunk 3]
```diff
@@ -152,3 +152,15 @@ def plot_link_up_down_count_per_aggregation_time(self):
         self._save_pivot_data_in_bars(
             pivot_links_data, "Time", "Number of Events", "Link up/down events", "Event"
         )
+
+    def full_analysis(self):
+        images, _, _ = super().full_analysis()
+        critical_events_headers = ["timestamp", "event_type", "event", "count"]
+        lists = [
+            (
+                self.get_critical_event_bursts(),
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py:161]: get_critical_event_bursts returns list only since it was implemented before the PDF dataframe support.
If you think you can do it now, you can change it to return a data frame. If not, let me know and I'll open a ticket to fix it when we can

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py [hunk 4]
```diff
@@ -156,7 +158,7 @@ def full_analysis(self):
                 except:  # pylint: disable=bare-except
                     pass
 
-        return self._images_created if len(self._images_created) > 0 else []
+        return self._images_created if len(self._images_created) > 0 else [], [], []
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py:161]: This change means that now, all class who would like to add a data frame or a text, will need to overwride this method.

We should have done this change in a way the full analysis of the base returns all the images, data frame and texts.

If the analysis function of the images, handles adding the images to the images list, we can do the same for data frames and the texts.
This way, we won't need overrides in the analyzers, "just" to make sure that each analysis function adds the data to the right location in the class

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py [hunk 5]
```diff
@@ -142,5 +142,7 @@ def plot_link_flapping_last_week(self):
         )
 
     def full_analysis(self):
-        self.get_link_flapping_last_week()
-        return super().full_analysis()
+        link_flapping_last_week = self.get_link_flapping_last_week()
+        images, _, _ = super().full_analysis()
+        df = [("Link Flapping last week", link_flapping_last_week,)]
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py]: Why the last comma in the tuple?

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ufm_top_analyzer.py [hunk 6]
```diff
@@ -13,20 +13,33 @@
...
-        return graphs_and_titles
+            tmp_images_list, tmp_dataframes, tmp_lists = analyzer.full_analysis()
+            graphs_and_titles.extend(tmp_images_list)
+            dataframes.extend(tmp_dataframes)
+            lists_to_add.extend(tmp_lists)
+
+        has_ibdiagnet_analyzer = any(isinstance(instance, IBDIAGNETLogAnalyzer) \
+                                     for instance in self._analyzers)
+        if not has_ibdiagnet_analyzer:
+            dataframes.append(("Fabric info", ("No Fabric Info found")))
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ufm_top_analyzer.py]: This should not be in this class; top analyzer should know any analyzer type.

I adding the "No Fabric Info found" text is an issue, I'm ok with not adding it at all.

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py [hunk 7]
```diff
@@ -369,7 +369,10 @@ def create_analyzer(
         end = time.perf_counter()
         log.LOGGER.debug(f"Took {end-start:.3f} to load the parsed data")
 
-        all_images_outputs_and_title = ufm_top_analyzer.full_analysis()
+        all_images_outputs_and_title, dataframes_for_pdf, lists_to_add = (
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py:372]: A general comment about the PR description:
How - I don't think we should have a link to Redmine here. This is a public repo, and this link does not help anyone from outside trying to understand things. 
Test- Please give more information on how it was tested

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py (hunk 1 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py (hunk 2 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py (hunk 4 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py]: I made changes to the function; it no longer returns anything. Instead, I added getters to retrieve images, data frames, or text separately.

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py (hunk 5 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ufm_top_analyzer.py (hunk 6 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ufm_top_analyzer.py]: deleted

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py (hunk 7 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: Test section was changed

Comment (boazhaim) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: and the why?

Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: I think it's useful and convenient that the link is there.

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py (hunk 3 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py]: done

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py (hunk 7 above)
Comment (boazhaim) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: Please remove it....
1. Not everyone who has access to the code has access to the ticket.
2. The PR itself should be clear in a way that there is no need to go to other links.
3. The ticket might be changed/deleted - what you write here will stay with the code.

If we move to another ticket system in two years and all the old tickets are lost, a reader will not have the full context of the code.

**Review by drorlevy — APPROVED** [review]

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py [hunk 8]
```diff
@@ -388,69 +391,9 @@ def create_analyzer(
...
-            lists_to_add.append(
-                (
-                    [cur_telemetry.get_number_of_core_dumps()],
-                    f"{cur_telemetry.telemetry_type} "
-                    "number of core dumps found in the logs",
-                    ["Amount"],
-                )
-            )
 
         # PDF creator gets all the images and to add to the report
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py:395]: I think this comment can be removed or updated

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py [hunk 9]
```diff
@@ -194,11 +205,14 @@ def __init__(
...
-            temp_file, "w", newline="", encoding=DataConstants.UTF8ENCODING
-        ) as outfile:
+        with (
+            open(
+                input_file, "r", newline="", encoding=DataConstants.UTF8ENCODING
+            ) as infile,
+            open(
+                temp_file, "w", newline="", encoding=DataConstants.UTF8ENCODING
+            ) as outfile,
+        ):
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py:215]: Why did you change it? I'm not sure I see the benefit

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py [hunk 10]
```diff
@@ -152,3 +147,11 @@ def plot_link_up_down_count_per_aggregation_time(self):
         self._save_pivot_data_in_bars(
             pivot_links_data, "Time", "Number of Events", "Link up/down events", "Event"
         )
+
+    def full_analysis(self):
+        super().full_analysis()
+        df_to_add = (
+            "More than 5 events burst over a minute",
+            self.get_critical_event_bursts(),
```
Comment (boazhaim) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py:155]: Why does it need to be here, why not like the images? where each analyzing function is run autmaticly and sets the output....

For example. in this case, get_critical_event_bursts once called, will add the data to the `dataframes_for_pdf`

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py [hunk 11]
```diff
@@ -193,3 +193,34 @@ def get_number_of_core_dumps(self):
             self._log_data_sorted["type"] == "timeout_dump_core"
         ]
         return {"Amount": len(core_dumps)}
+
+    def full_analysis(self):
```
Comment (TalLerner) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py:197]: Miryam,
You misunderstood the purpose of the assignment. The base class already created an infrastructure to run the analysis. You should add methods to run as part of the base class full_analysis and NOT implement your own full_analysis method.
If it not clear let talk.

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet_log_analyzer.py [hunk 12]
```diff
@@ -33,4 +33,5 @@ def full_analysis(self):
         Returns a list of all the graphs created and their title
         """
         self.print_fabric_size()
-        return []
+        fabric_info = self.get_fabric_size()
```
Comment (TalLerner) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet_log_analyzer.py:36]: Same thing. Use the infrastructure in the base class to run this.

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py [hunk 13]
```diff
@@ -142,5 +142,7 @@ def plot_link_flapping_last_week(self):
         )
 
     def full_analysis(self):
-        self.get_link_flapping_last_week()
-        return super().full_analysis()
+        super().full_analysis()
```
Comment (TalLerner) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py:145]: Same thing. Use the infrastructure in the base class to run this.

**Review by TalLerner — CHANGES_REQUESTED** [review]

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/pdf_creator.py [hunk 14]
```diff
@@ -85,7 +86,7 @@ def add_list_of_dicts_as_text(self, data_list, title=None, headers=None):
 
     def add_dataframe_as_text(self, data_frame, title=None):
         """Adds a DataFrame to the PDF as aligned text without row numbers."""
-        if data_frame is None or data_frame.empty:
+        if not isinstance(data_frame, pd.DataFrame) or data_frame.empty:
```
[the remark this answers is not in the document]
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/pdf_creator.py:89]: if data_frame is None, 'isinstance(data_frame, pd.DataFrame)' will return False

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py (hunk 7 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: I removed it

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py (hunk 9 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py:215]: ruff requires me to change

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/console_log_analyzer.py [hunk 15]
```diff
@@ -108,4 +111,4 @@ def full_analysis(self):
         Returns a list of all the graphs created and their title
         """
         self.print_exceptions()
-        return super().full_analysis()
+        super().full_analysis()
```
[the remark this answers is not in the document]
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/console_log_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py [hunk 16]
```diff
@@ -193,3 +193,34 @@ def get_number_of_core_dumps(self):
...
+        super().full_analysis()
+        self._dataframes_for_pdf.append(
+            (
+                f"{self.telemetry_type} Telemetry iteration time",
```
[the remark this answers is not in the document]
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py [hunk 17]
```diff
@@ -193,3 +193,34 @@ def get_number_of_core_dumps(self):
...
+                self.get_last_iterations_time_stats(),
+            )
+        )
+
+        self._dataframes_for_pdf.append(
+            (
+                f"{self.telemetry_type} "
+                "Telemetry iteration first and last timestamps",
```
[the remark this answers is not in the document]
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py [hunk 18]
```diff
@@ -193,3 +193,34 @@ def get_number_of_core_dumps(self):
...
+            (
+                f"{self.telemetry_type} Telemetry fabric size",
+                self.get_number_of_switches_and_ports(),
+            )
+        )
+
+        self._txt_for_pdf.append(
+            (
+                [self.get_number_of_core_dumps()],
+                f"{self.telemetry_type} " "number of core dumps found in the logs",
```
[the remark this answers is not in the document]
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py [hunk 19]
```diff
@@ -193,3 +193,34 @@ def get_number_of_core_dumps(self):
...
+        self._dataframes_for_pdf.append(
+            (
+                f"{self.telemetry_type} "
+                "Telemetry iteration first and last timestamps",
+                self.get_first_last_iteration_timestamp(),
+            )
+        )
+        self._dataframes_for_pdf.append(
+            (
+                f"{self.telemetry_type} Telemetry fabric size",
```
[the remark this answers is not in the document]
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py (hunk 8 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzer.py]: removed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py (hunk 10 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/events_log_analyzer.py:161]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py (hunk 11 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet2_port_counters_analyzer.py]: OK. I changed it

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet_log_analyzer.py (hunk 12 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/ibdiagnet_log_analyzer.py:33]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py (hunk 13 above)
Comment (Miryam-Schwartz) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/link_flapping_analyzer.py]: fixed

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py [hunk 20]
```diff
@@ -194,11 +205,14 @@ def __init__(
     def _remove_empty_lines_from_csv(input_file):
         temp_file = input_file + ".temp"
 
-        with open(
-            input_file, "r", newline="", encoding=DataConstants.UTF8ENCODING
-        ) as infile, open(
-            temp_file, "w", newline="", encoding=DataConstants.UTF8ENCODING
-        ) as outfile:
+        with (
```
Comment (TalLerner) [inline plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py:208]: Please revert this display

**Review by TalLerner — CHANGES_REQUESTED** [review]

File: plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py (hunk 9 above)
Comment (boazhaim) [reply plugins/ufm_log_analyzer_plugin/src/loganalyze/log_analyzers/base_analyzer.py:215]: I tried to fix it, really strange.... lets keep this change.
I pushed some fixes to the ruff step

**Review by boazhaim — APPROVED** [review]
